### PR TITLE
Better rendering of the YouTube feed content.

### PR DESCRIPTION
**Pull request by javerous**
Hello,

Here is a change proposal that I'm using and that I think can be interesting to include in the main project.

The global idea is that it was interesting to me to add the original YouTube feed content in articles (in addition to the iframe), as it includes some description / comment from the author, but the FreshRSS generic rendering of this content was not good enough:
- The description uses a `<pre>` node, which is a bit "raw".
- Thumbnail is visible, so it's a bit duplication with the preview of the YouTube player we can see in the iframe.

So I added a bit more evolved formatting:
- The description is now in a `<p>` node, so it shows as standard text.
- Thumbnail is added, but **hidden**, this way RSS clients (like mine) can still extract it to show a content preview, but then it's not redundant in the article itself with the iframe.
- The title is hidden, as RSS clients already know it (but still included, so if some RSS client extract it, it's still available).

I wanted to keep the choice of feed content inclusion, so I converted the original "show-content" checkbox to a group of radio buttons, so the user can choose between "None" (i.e doesn't include original feed content, i.e. show only the iframe), "Append" (i.e. iframe + content generated by FreshRSS) and "Format" (i.e. iframe + reformatted original feed content).

It's possible that naming of the 3 modes have to be enhanced…

On a side point, I changed a bit configuration handling:
- The code was in duplication with what was already done in `extension.php`, which was a bit annoying for my change.
- The `empty(FreshRSS_Context::$user_conf->yt_xxx)` was not working on my side, for some reason (always returning true, so never actually reading settings). Not sure what is the point, but I changed to a simple `!= ''` comparison.

Note:
- I'm using `nl2br(htmlentities($descriptions[0]->nodeValue))`, because, as an YouTube extension, we know that description will be always pure text, and not pre-formated HTML (well, except if YouTube decide one day to give pre-formated HTML content…).

I'm not a PHP developer, so I might have didn't do it well.

If you are not interested in this change, feel free to reject it, I will keep it for myself ;)

Finally, some screen-shot to show the 3 modes in my RSS client (Reeder):
- "None":
yt_none
 
- "Append":
yt_append

- "Format":
yt_format

### CODE REVIEW

File: xExtension-YouTube/extension.php
Comment (kevinpapst) [inline xExtension-YouTube/extension.php]: Is `<p hidden>` proper HTML? Should it be `<p class="hidden">`?

File: xExtension-YouTube/i18n/en/ext.php
Comment (kevinpapst) [inline xExtension-YouTube/i18n/en/ext.php]: Can we describe these a little better? I know this will probably not seen often by a user, but from just reading these titles I don't understand what it means.

File: xExtension-YouTube/extension.php
Comment (kevinpapst) [inline xExtension-YouTube/extension.php]: Nitpicking ... but I have never seen a method that is called `forgeX` - should that be `formatX`?

File: xExtension-YouTube/configure.phtml
Comment (kevinpapst) [inline xExtension-YouTube/configure.phtml]: lets use `===` instead of `==` and for the other lines as well

File: xExtension-YouTube/extension.php
Comment (kevinpapst) [inline xExtension-YouTube/extension.php]: I'd say: lets make that the "default" case.

File: xExtension-YouTube/extension.php
Comment (kevinpapst) [inline xExtension-YouTube/extension.php]: I like that there is a smooth upgrade path for previous users!

**Review by kevinpapst — CHANGES_REQUESTED** [review]
A nice addition which I really like. Thanks for creating the PR @javerous 👍 

There are some smaller things I'd like to discuss and I'd like to invite @marienfressinaud to the PR, as I  currently don't have a testing instance for FRSS.

File: xExtension-YouTube/extension.php
Comment (javerous) [reply xExtension-YouTube/extension.php]: It's what I saw there… https://www.w3schools.com/tags/att_hidden.asp

That being said, I'm not an HTML guru, so it was perhaps the old way to do it ? At least, it seems that WebKit (the mac / iOS HTML renderer) recognize it 🤷‍♂️

Comment (kevinpapst) [reply xExtension-YouTube/extension.php]: Interesting, I have never seen that in the wild... thanks for teaching me something new

File: xExtension-YouTube/extension.php
Comment (javerous) [reply xExtension-YouTube/extension.php]: Or `getHtml` ? The idea was mainly to remove the `iframe` from the name, as it can contain other things.

File: xExtension-YouTube/extension.php
Comment (javerous) [reply xExtension-YouTube/extension.php]: Done.

File: xExtension-YouTube/configure.phtml
Comment (javerous) [reply xExtension-YouTube/configure.phtml]: Done.

File: xExtension-YouTube/extension.php
Comment (javerous) [reply xExtension-YouTube/extension.php]: Done.

File: xExtension-YouTube/i18n/en/ext.php
Comment (javerous) [reply xExtension-YouTube/i18n/en/ext.php]: I need to think about it, I'm not very inspired ;p

Comment (javerous) [reply xExtension-YouTube/i18n/en/ext.php]: Perhaps:
```
'title' => 'Content'
'none' => 'Player Only'
'append' => 'Player + Raw Feed'
'format' => 'Player + Formatted Feed'
```
? It has the advantage to be compact and somehow factual…

Comment (kevinpapst) [reply xExtension-YouTube/i18n/en/ext.php]: LOL, what about something like this:
```
title = How to handle the feed content
none => Include the video only
append => Add hidden content for external feed readers
format => Add formatted content 
```
Or was it (append<=>format) vice versa?
Just that people understand via there are more options...

Update: seems you were quicker^^ better, but I would still include the intention (the external feed reader)

Comment (javerous) [reply xExtension-YouTube/i18n/en/ext.php]: For "append" mode, it's stricly the same than previously when 'Display the feeds content' was checked: it include the HTML content generated by FreshRSS (which is essentially a div with an `enclosure` class, and some sub-elements with their dedicated classes) just after the YouTube player (the iframe).

This content is recognized (and extracted) by some RSS client (at least mine), but it stay in the article view anyway. It's why we see the YouTube player (iframe) + (a part) of the thumbnail in the "append" screenshot (the FreshRSS generated content).

The interest of the "format" mode (vs the "append" mode) is that it include the same classes, but either hidden (for thumbnail & title), either reformated (for the description).

I'm open to even rename the modes (in addition of the titles) if it's not clear.

Comment (javerous) [reply xExtension-YouTube/i18n/en/ext.php]: To make it more clear, the point is about the "hidden" word in your suggestion. It's true that title & thumbnail are hidden, but it's not the case of the description. RSS client (at least mine) doesn't extract it, it just display it "as it".

With the FreshRSS generated content (using a `<pre>` node), it looks like this:
Screenshot 2020-01-12 at 21 24 15

With this PR formated content (using a `<p>` node), it looks like this:
Screenshot 2020-01-12 at 21 24 54

Which, from my point of view, look better.

Hiding in some parts, and changing description format change the global look & feel, i.e. the global format, so the "format" mode ^^'

**Comment by kevinpapst** [comment]
@marienfressinaud any chance you do a review & test here? I still don't have a running FRSS instance and no time currently to set it up.

**Comment by marienfressinaud** [comment]
Yes I'll do, but more probably on Thursday ;)

File: xExtension-YouTube/configure.phtml
Comment (marienfressinaud) [inline xExtension-YouTube/configure.phtml]: This part can be improved by placing the input and label inside a `<label>` tag so all the lines become clickable. Note I'm not sure it's good practice to place input in a label, but it's how it's done for checkboxes and radios elsewhere in FreshRSS code so let's stick to the standard :). I suggest:

```html
		<label class="group-name" for="yt_content_none"><?php echo _t('ext.yt_videos.content.title'); ?></label>
		<div class="group-controls">
			<label for="yt_content_none">
				<input type="radio" id="yt_content_none" name="yt_content_choice" value="none" <?= $this->getContentHandling() === 'none' ? 'checked' : '' ?>>
				<?php echo _t('ext.yt_videos.content.none'); ?>
			</label>
			<label for="yt_content_append">
				<input type="radio" id="yt_content_append" name="yt_content_choice" value="append" <?= $this->getContentHandling() === 'append' ? 'checked' : '' ?>>
				<?php echo _t('ext.yt_videos.content.append'); ?>
			</label>
			<label for="yt_content_format">
				<input type="radio" id="yt_content_format" name="yt_content_choice" value="format" <?= $this->getContentHandling() === 'format' ? 'checked' : '' ?>>
				<?php echo _t('ext.yt_videos.content.format'); ?>
			</label>
		</div>
```

File: xExtension-YouTube/extension.php
Comment (marienfressinaud) [inline xExtension-YouTube/extension.php:72]: You're reintroducing [a bug](https://github.com/kevinpapst/freshrss-youtube/pull/13) when user hasn't configured the extension yet. Alternative can be:

```php
if (isset(FreshRSS_Context::$user_conf->yt_player_width) && FreshRSS_Context::$user_conf->yt_player_width != '') {
    …
}
```

You'll have to take care of that too for the option you're adding.

File: xExtension-YouTube/configure.phtml
Comment (marienfressinaud) [inline xExtension-YouTube/configure.phtml]: I think it should be placed elsewhere… Maybe in the `init` method of the extension.php file?

File: xExtension-YouTube/i18n/fr/ext.php
Comment (marienfressinaud) [inline xExtension-YouTube/i18n/fr/ext.php]: (nitpick) indentation

File: xExtension-YouTube/extension.php
Comment (marienfressinaud) [inline xExtension-YouTube/extension.php]: Maybe the default value (second parameter of `param`) should be `none` or other default value of `contentHandling`.

(nitpick) You don't need to cast a string in string ;)

File: xExtension-YouTube/extension.php
Comment (marienfressinaud) [inline xExtension-YouTube/extension.php]: (nitpick) What's the purpose of the `[error]` value? `$content` seems to be always overridden, isn't it?

File: xExtension-YouTube/extension.php
Comment (marienfressinaud) [inline xExtension-YouTube/extension.php]: Maybe a comment in the code can help next developers to understand why it's important to add hidden elements? :)

**Review by marienfressinaud — CHANGES_REQUESTED** [review]
I wonder if it wouldn't be better to remove the `contentHandling` option and keep only the `format` behaviour? It's a really nice improvement and I don't see the point to keep the 2 other options

Enclosures should have a better look by default, I don't understand why `<pre>` were used at first :man_shrugging:

File: xExtension-YouTube/configure.phtml
Comment (javerous) [reply xExtension-YouTube/configure.phtml]: Yes, good point. Fixed (and tested).

File: xExtension-YouTube/configure.phtml
Comment (javerous) [reply xExtension-YouTube/configure.phtml]: Yes, I agree. The init, or having the getters be lazy, and load their values when executed the first time ?

Not sure it was not done in the init in the first place. Perhaps too early? I will test.

File: xExtension-YouTube/i18n/fr/ext.php
Comment (javerous) [reply xExtension-YouTube/i18n/fr/ext.php]: Fixed ;)

File: xExtension-YouTube/extension.php
Comment (javerous) [reply xExtension-YouTube/extension.php]: Yes, I changed this fact by @kevinpapst suggestion, and forgot to remove this. By the way, what is the practice in this case, in PHP ?

`$content = '';` and let the switch cases override it, or just not define it, and let the switch cases defining it ?

Comment (marienfressinaud) [reply xExtension-YouTube/extension.php]: I personally prefer to define it with `$content = '';` so I'm sure the variable exists after the `switch`, but I don't know if there's any good PHP practice that defines this ^^

File: xExtension-YouTube/extension.php
Comment (javerous) [reply xExtension-YouTube/extension.php]: Yes ! Done.

File: xExtension-YouTube/extension.php
Comment (kevinpapst) [reply xExtension-YouTube/extension.php]: Even though it is not really necessary in PHP I would also always define it above the switch, either as null or as empty string if required. The `[error]` will never be used, so it should be changed.

File: xExtension-YouTube/i18n/en/ext.php
Comment (kevinpapst) [inline xExtension-YouTube/i18n/en/ext.php:6]: I know, this one is annoying ;-) but it has to be changed. 
And it is not up to discussion, because every developer has to decide that one for himself ... all of my projects use 4 spaces for one tab.

File: xExtension-YouTube/configure.phtml
Comment (kevinpapst) [reply xExtension-YouTube/configure.phtml]: Is that even required? I don't think so.

**Comment by javerous** [comment]
> I wonder if it wouldn't be better to remove the `contentHandling` option and keep only the `format` behaviour? It's a really nice improvement and I don't see the point to keep the 2 other options

Well, I was wondering, but I wanted to have my change be the least intrusive possible, so let the choice between the previous possiblies, and add this new one. Perhaps some people would prefer to have the player only (for some reason) ? I don't know. I'm open to remove the setting, if you feel we should stay always with the "format" mode.

> Enclosures should have a better look by default, I don't understand why `<pre>` were used at first 🤷‍♂

I was wondering too, but my guess is because you don't know in which format the content is going to be. If it's HTML, then it would be okay to include it directly (even if probably ugly in a <pref>), but if it's raw text, you have to HTML-ize it a bit (at least for new lines), but then you would create a mess for HTML content. And I guess it's a bit hard to detect that. For this extension, we know it will never (?) be in HTML, as we control on which kind of feed it apply.

I'm not sure, but yes, this `<pre>` is ugly.

File: xExtension-YouTube/extension.php
Comment (javerous) [reply xExtension-YouTube/extension.php:72]: I just tried, but `isset` is always false on my case, even when there is actual values (probably related to the `empty` which was always true, which is why I removed them), so it force values to be on default… :/

Any ideas why ?

File: xExtension-YouTube/configure.phtml
Comment (javerous) [reply xExtension-YouTube/configure.phtml]: To load them ? Yes, for the UI, as I converted `FreshRSS_Context::$user_conf->xxx` to the class getter (it was to simplify the code, because of the old confing conversion, to prevent code duplication).

**Comment by kevinpapst** [comment]
I am one of the people who want to see the player only ;-) I don't care about the content for these feeds. The third option could be removed for my personal use-case. Normally having less options is always great for the human brain and its less code that needs to be taken care of.

File: xExtension-YouTube/extension.php
Comment (javerous) [reply xExtension-YouTube/extension.php]: Ah yes, indeed, I wasn't sure of the actual return type.

For the default, my idea was to delegate that to settings loading in `loadConfigValues()`, so it's all in the same place. A bit like for height & width.

File: xExtension-YouTube/extension.php
Comment (javerous) [reply xExtension-YouTube/extension.php]: Okay, fixed.

File: xExtension-YouTube/i18n/en/ext.php
Comment (javerous) [reply xExtension-YouTube/i18n/en/ext.php:6]: You want me to use 4 spaces instead of tabs ? It's okay, but just for my own defense (:p) those files were already mixing tabs & spaces and configure.ptml use tabs all around ^^'

File: xExtension-YouTube/extension.php
Comment (marienfressinaud) [reply xExtension-YouTube/extension.php:72]: I'll take a look after lunch ;)

File: xExtension-YouTube/i18n/en/ext.php
Comment (kevinpapst) [reply xExtension-YouTube/i18n/en/ext.php:6]: Wow, you are right. I can't believe I committed tabs. Keep it then as is. If more files have to be changed, its better to have all format changes in one single commit.

File: xExtension-YouTube/extension.php
Comment (marienfressinaud) [reply xExtension-YouTube/extension.php]: Ok then :+1:

**Review by marienfressinaud — COMMENTED** [review]
Let's keep an option to toggle content / no contents then :+1: Third option (i.e. append) looks almost counter-productive with a degraded experience so I would remove it

File: xExtension-YouTube/configure.phtml
Comment (javerous) [reply xExtension-YouTube/configure.phtml]: Okay, I tested in the `init`, but the settings in UI are "out of phase" (we always see the previous setting). I have the feeling they don't call the `init()` at all for the configuration UI (and that it's the call in the `handleConfigureAction` which make them appear anyway).

Not sure why `loadConfigValues` is called in `handleConfigureAction` by the way.

File: xExtension-YouTube/extension.php
Comment (marienfressinaud) [reply xExtension-YouTube/extension.php:72]: I can't reproduce the error I had the other time, really strange :disappointed: so let's keep it like this and sorry for the noise!

File: xExtension-YouTube/configure.phtml
Comment (marienfressinaud) [reply xExtension-YouTube/configure.phtml]: Ah right: `handleConfigureAction` is called both when showing the panel and when validating the form. So the call to `loadConfigValues` in this method is enough. However, it must be placed after saving the new values so it synchronizes internal values. In consequence, you just have to move `$this->loadConfigValues();` at the end of `handleConfigureAction` and you should be fine :+1:

File: xExtension-YouTube/extension.php
Comment (javerous) [reply xExtension-YouTube/extension.php:72]: Perhaps some updates in PHP ? Or changes in the way FreshRSS expose those contexts ?

File: xExtension-YouTube/configure.phtml
Comment (javerous) [reply xExtension-YouTube/configure.phtml]: Fixed, thanks !

**Comment by javerous** [comment]
Okay, I think I handled all code review. I will switch back to a checkbox to just enable content (with better formatting), or no content, as it was before.

File: xExtension-YouTube/extension.php
Comment (marienfressinaud) [reply xExtension-YouTube/extension.php:72]: Nope I don't think so. It's more probably I was doing some other things and the errors were caused by my own code :thinking:

**Comment by javerous** [comment]
@marienfressinaud Here we are. I put back the check box. I tested, everything seems to work.

File: xExtension-YouTube/i18n/fr/ext.php
Comment (marienfressinaud) [inline xExtension-YouTube/i18n/fr/ext.php]:
```suggestion
		'use_nocookie' => 'Utiliser le domaine www.youtube-nocookie.com pour éviter les cookies',
```

File: xExtension-YouTube/extension.php
Comment (marienfressinaud) [inline xExtension-YouTube/extension.php]: (nitpick) `isShowContent` was fine to me (but I don't really care)

File: xExtension-YouTube/configure.phtml
Comment (marienfressinaud) [inline xExtension-YouTube/configure.phtml:27]: If you still have some motivation, you can improve the look of the checkboxes by moving the labels on the right:

```html
		<div class="group-controls">
			<label class="checkbox" for="yt_show_content">
				<input type="checkbox" id="yt_show_content" name="yt_show_content" value="1" <?= $this->getShowContent() ? 'checked' : '' ?>>
				<?= _t('ext.yt_videos.show_content') ?>
			</label>
		</div>

		<div class="group-controls">
			<label class="checkbox" for="yt_nocookie">
				<input type="checkbox" id="yt_nocookie" name="yt_nocookie" value="1" <?= $this->getUseNoCookie() ? 'checked' : '' ?>>
				<?= _t('ext.yt_videos.use_nocookie') ?>
			</label>
		</div>
```

It looks like this 

Capture d’écran de 2020-01-17 09-17-09

**Review by marienfressinaud — APPROVED** [review]
My last comments, nothing critical so feel free to let them for another PR :)

File: xExtension-YouTube/configure.phtml
Comment (javerous) [reply xExtension-YouTube/configure.phtml:27]: Done ;)

File: xExtension-YouTube/extension.php
Comment (javerous) [reply xExtension-YouTube/extension.php]: Done.

File: xExtension-YouTube/i18n/fr/ext.php
Comment (javerous) [reply xExtension-YouTube/i18n/fr/ext.php]: Yes, I wanted to translate, but as it was not part of this PR, so…

It's done 👍

**Comment by marienfressinaud** [comment]
Perfect :ok_hand: just have to wait for @kevinpapst to review and merge now :)

File: xExtension-YouTube/configure.phtml
Comment (kevinpapst) [inline xExtension-YouTube/configure.phtml]: Event though short echo tags are officially first class citizens in PHP since 5.4 (and we can safely argue that we give a shit on users with 5.3), I don't like them ([no matter what PSR-1 says](https://www.php-fig.org/psr/psr-1/#21-php-tags)) - but even more important: a PR should not change formattings or method names or code styles without good reasons. It makes the life of the reviewer so much more complicated.
As `<?php` vs `<?=` is a semi-religious discussion (like tabs vs spaces) please revert them.

File: xExtension-YouTube/configure.phtml
Comment (kevinpapst) [inline xExtension-YouTube/configure.phtml]: @marienfressinaud do you think this is a safe decision to use `this` here for the future? Feels like "works by accident" or was this your intention originally when adding plugin support?

File: xExtension-YouTube/extension.php
Comment (kevinpapst) [inline xExtension-YouTube/extension.php]: Lets don't mix `isser` and `getter` for bools => `isUseNoCookieDomain()`

File: xExtension-YouTube/configure.phtml
Comment (marienfressinaud) [reply xExtension-YouTube/configure.phtml]: Can't remember what was my first intention, but at least I used it in the [CustomCSS extension](https://github.com/FreshRSS/Extensions/blob/master/xExtension-CustomCSS/configure.phtml). So if it wasn't officially supported, let's say it is now :D

Comment (kevinpapst) [reply xExtension-YouTube/configure.phtml]: Thats a word 😁 resolved

File: xExtension-YouTube/configure.phtml
Comment (marienfressinaud) [reply xExtension-YouTube/configure.phtml]: Just a note to say that FRSS now supports only from PHP 5.6 and above, and use short tags to reduce the amount of `echo` in the code since the last version. That being said, it's your extension so your choices of course :)

Comment (kevinpapst) [reply xExtension-YouTube/configure.phtml]: Thanks for the feedback! I haven't worked on a project that is using PHP templates for a long time, so maybe this is more state of the art nowadays. Maybe it is just me, using PHP for almost 2 decades now and having a lot of troubles with them in the past ... old habits are very strong. But as I said "semi-religious discussion" and those are vey rarely driven by logic ^^ So this is one of the points where I will shamelessly use my repository owner decision power 😈

**Comment by kevinpapst** [comment]
btw @javerous - sorry for the long review and thanks for staying with us during that hard time 😀

**Comment by javerous** [comment]
@kevinpapst Ha ha, no problem, it was interesting. I'm not a really a PHP developer, so I learnt a bunch of thing ;)

File: xExtension-YouTube/configure.phtml
Comment (javerous) [reply xExtension-YouTube/configure.phtml]: Done.

**Review by kevinpapst — APPROVED** [review]

**Comment by kevinpapst** [comment]
@javerous as there is not even a linter running in this old repo: if you tested your latest changes I will hit the merge button?!

**Comment by javerous** [comment]
I did basic tests, everything is working yes (the UI appear, and button ares working and their states well saved) !

**Comment by kevinpapst** [comment]
Perfect, then once again: thanks for your contribution 🚀  available as new release  https://github.com/kevinpapst/freshrss-youtube/releases/tag/0.10

Last thing that needs to be done: open a PR ion the extensions repo to change the release version here: https://github.com/FreshRSS/Extensions/blob/master/extensions.json#L39

**Comment by javerous** [comment]
@kevinpapst You didn't bump the version in the meta file and in the readme ;)

**Comment by kevinpapst** [comment]
True ...  hm, well - if you want to do it, please go ahead :D

**Pull request merged**
